### ISSUE TITLE
AST: Store a Single Parameter in TSIndexSignature

### ISSUE THREAD

**Issue by camc314** [oxc-project/oxc#25684]
Parent: oxc-project/backlog#210

## Summary

Replace `TSIndexSignature::parameters: Vec<TSIndexSignatureName>` with one required `parameter: TSIndexSignatureName`.

TypeScript index signatures require exactly one parameter. The vector mirrors ESTree's external array shape rather than the TypeScript grammar and permits invalid zero- and multi-parameter index signatures in the Rust AST.

## Motivation

A valid TypeScript index signature has exactly one parameter:

```ts
type MapOf<T> = {
  [key: string]: T;
};
```

The current AST stores that parameter in a vector:

```rs
pub struct TSIndexSignature<'a> {
    pub node_id: Cell<NodeId>,
    pub span: Span,
    pub parameters: Vec<'a, TSIndexSignatureName<'a>>,
    pub type_annotation: Box<'a, TSTypeAnnotation<'a>>,
    pub readonly: bool,
    pub r#static: bool,
}
```

This makes the following invalid states representable:

```text
parameters.len() == 0
parameters.len() > 1
```

There is no valid TypeScript syntax for either state:

```ts
type Missing = {
  []: unknown;
};

type Multiple = {
  [first: string, second: number]: unknown;
};
```

The parser already enforces the real invariant after parsing a comma-delimited list:

```rs
match params.as_slice() {
    [param] => {
        // Validate the parameter type.
    }
    _ => self.error(diagnostics::index_signature_one_parameter(...)),
}
```

The invalid list is nevertheless stored in `TSIndexSignature`, so downstream consumers must remain defensive. Codegen loops over every parameter and prints commas, while linter rules use `parameters.first()` and provide a fallback for an absent parameter.

The vector also adds a length/capacity header and a separate arena allocation for a value whose valid cardinality is statically one.

## Current AST Shape

```rs
pub struct TSIndexSignature<'a> {
    pub node_id: Cell<NodeId>,
    pub span: Span,
    pub parameters: Vec<'a, TSIndexSignatureName<'a>>,
    pub type_annotation: Box<'a, TSTypeAnnotation<'a>>,
    pub readonly: bool,
    pub r#static: bool,
}

pub struct TSIndexSignatureName<'a> {
    pub node_id: Cell<NodeId>,
    pub span: Span,
    #[estree(json_safe)]
    pub name: Str<'a>,
    pub type_annotation: Box<'a, TSTypeAnnotation<'a>>,
}
```

`TSIndexSignatureName` serializes as an ESTree `Identifier` containing its type annotation.

## Proposed AST Shape

Store the grammar-required parameter directly:

```rs
pub struct TSIndexSignature<'a> {
    pub node_id: Cell<NodeId>,
    pub span: Span,
    #[estree(rename = "parameters", via = TSIndexSignatureParameters)]
    pub parameter: TSIndexSignatureName<'a>,
    pub type_annotation: Box<'a, TSTypeAnnotation<'a>>,
    pub readonly: bool,
    pub r#static: bool,
}
```

`TSIndexSignatureParameters` should serialize the singular Rust parameter as a one-element ESTree
array and extract the single element during raw transfer.

The parameter is shown inline to remove the vector allocation. The final inline-versus-boxed layout
should be selected using the generated size and allocation checks.

## Syntax Mapping

| Source form | Proposed representation |
| --- | --- |
| `{ [key: string]: Value }` | `parameter: TSIndexSignatureName { name: "key", .. }` |
| `{ readonly [index: number]: Value }` | One parameter with `readonly: true` |
| `class C { static [key: string]: Value }` | One parameter with `static: true` |
| `{ [a: string, b: number]: Value }` | Invalid input; not representable as a valid `TSIndexSignature` |
| `{ []: Value }` | Invalid input; not representable as a valid `TSIndexSignature` |

## Guaranteed Invariants

- Every `TSIndexSignature` contains exactly one parameter.
- Consumers do not need length checks, iteration, `.first()`, or an absent-parameter fallback.
- Codegen cannot emit a zero- or multi-parameter index signature from a valid Rust AST node.
- The parameter name and its type annotation remain grouped in `TSIndexSignatureName`.

**Comment by overlookmotel** [comment]
Nice! Current AST shape makes no sense at all. This is a definite improvement.

One small thing: `TSIndexSignatureName::name` field should probably be an `Ident`, not a `Str`. Could also then remove `#[estree(json_safe)]` from that field, as `Ident`s are always JSON-safe.
